### PR TITLE
Issue 4592: dscreate error with custom dir_path

### LINKED ISSUE #4592: dscreate error, cannot start ns-slapd with custom db_dir path, incorrect sub dir permissions
Description of problem:

dscreate fails when using custom path with the db_dir parameter.
Creating the db_dir path and changing permissions, mode, SELinux labels, prior to the dscreate did not seem to help.
It seems the directories are created in /usr/lib/python3.6/site-packages/lib389/instance/setup.py
The ns-slapd process start sequence fails with a permission error on the custom db directory.
CRIT - bdb_start - Can't start because the database directory "/testdata/dirsrv/test/db" either doesn't exist, or is not accessible

permissions and ownership look correct, SELinux labels not correct
have
unconfined_u:object_r:default_t:s0

versus an expected
unconfined_u:object_r:dirsrv_var_lib_t:s0

but permissive mode does not change the behavior.

the problem may be around line 757, when the parent directories owned by root:root are created with mode 770, while they need 775 so that the ns-slapd's uid/group can actually change into those directories:

/usr/lib/python3.6/site-packages/lib389/instance/setup.py
...
    def _install_ds(self, general, slapd, backends):
...
        # Create all the needed paths
        # we should only need to make bak_dir, cert_dir, config_dir, db_dir, ldif_dir, lock_dir, log_dir, run_dir?
        for path in ('backup_dir', 'cert_dir', 'db_dir', 'ldif_dir', 'lock_dir', 'log_dir', 'run_dir'):
            self.log.debug("ACTION: creating %s", slapd[path])
            try:
                os.umask(0o007)  # For parent dirs that get created -> sets 770 for perms
                os.makedirs(slapd[path], mode=0o770)
            except OSError:
                pass
            os.chown(slapd[path], slapd['user_uid'], slapd['group_gid'])

related:
os.makedirs(name, mode=0o777, exist_ok=False)¶
...
Changed in version 3.7: The mode argument no longer affects the file permission bits of newly-created intermediate-level directories.

workaround, do before the dscreate command:

mkdir -p /testdata/dirsrv/test/db
chmod 775 /testdata/
chmod 775 /testdata/dirsrv/
chown root:root /testdata/
chown root:root /testdata/dirsrv/
chmod -R 770 /testdata/dirsrv/test/
chown -R ldapuser1:ldapgroup1 /testdata/dirsrv/test/
chcon -R system_u:object_r:dirsrv_var_lib_t:s0 /testdata/dirsrv/
ls -alRZ /testdata/
/testdata/:
total 4
drwxr-xr-x.  3 root root unconfined_u:object_r:default_t:s0      20 Feb  2 19:55 .
dr-xr-xr-x. 20 root root system_u:object_r:root_t:s0           4096 Feb  2 19:55 ..
drwxr-xr-x.  3 root root system_u:object_r:dirsrv_var_lib_t:s0   18 Feb  2 19:55 dirsrv

/testdata/dirsrv:
total 0
drwxr-xr-x. 3 root      root       system_u:object_r:dirsrv_var_lib_t:s0 18 Feb  2 19:55 .
drwxr-xr-x. 3 root      root       unconfined_u:object_r:default_t:s0    20 Feb  2 19:55 ..
drwxrwx---. 3 ldapuser1 ldapgroup1 system_u:object_r:dirsrv_var_lib_t:s0 16 Feb  2 22:15 test

/testdata/dirsrv/test:
total 0
drwxrwx---. 3 ldapuser1 ldapgroup1 system_u:object_r:dirsrv_var_lib_t:s0 16 Feb  2 22:15 .
drwxr-xr-x. 3 root      root       system_u:object_r:dirsrv_var_lib_t:s0 18 Feb  2 19:55 ..
drwxrwx---. 2 ldapuser1 ldapgroup1 system_u:object_r:dirsrv_var_lib_t:s0  6 Feb  2 22:15 db

/testdata/dirsrv/test/db:
total 0
drwxrwx---. 2 ldapuser1 ldapgroup1 system_u:object_r:dirsrv_var_lib_t:s0  6 Feb  2 22:15 .
drwxrwx---. 3 ldapuser1 ldapgroup1 system_u:object_r:dirsrv_var_lib_t:s0 16 Feb  2 22:15 ..

Version-Release number of selected component (if applicable):

redhat-release-8.3-1.0.el8.x86_64
SELinux enforced or permissive
389-ds-base-1.4.3.13-1.module+el8dsrv+8334+69a46a2e.x86_64

How reproducible:
on demand

Steps to Reproduce:

1. create a config file

grep ldap /etc/passwd /etc/group
groupadd -r ldapgroup1
useradd -r -g  ldapgroup1 ldapuser1
grep ldap /etc/passwd /etc/group

cat << EOF > ~/ds11.instance.test.inf
[general]
full_machine_name = m2.example.test
[slapd]
backup_dir = /dump/dirsrv/{instance_name}
cert_dir = /etc/dirsrv/slapd-{instance_name}
config_dir = /etc/dirsrv/slapd-{instance_name}
data_dir = /usr/share
db_dir = /testdata/dirsrv/{instance_name}/db
user = ldapuser1
group = ldapgroup1
instance_name = test
port = 1389
secure_port = 1636
root_dn = cn=Directory Manager
root_password = password
[backend-userroot]
EOF

2. create an instance

dscreate -v from-file ~/ds11.instance.test.inf

3. review errors log and test

grep CRIT /var/log/dirsrv/slapd-test/errors

Actual results:

DEBUG: PASSED: instance checking
DEBUG: INFO: temp root password set to .PttHq45IET43tXes746XQ.RKPEP7WuzBsBCVdxCDoqmNcgZWJtuql1.1TqB7pxsY
DEBUG: PASSED: root user checking
DEBUG: PASSED: network avaliability checking
DEBUG: READY: Beginning installation for test...
DEBUG: ACTION: Creating dse.ldif
DEBUG: ACTION: creating /dump/dirsrv/test
DEBUG: ACTION: creating /etc/dirsrv/slapd-test
DEBUG: ACTION: creating /testdata/dirsrv/test/db
DEBUG: ACTION: creating /var/lib/dirsrv/slapd-test/ldif
DEBUG: ACTION: creating /var/lock/dirsrv/slapd-test
DEBUG: ACTION: creating /var/log/dirsrv/slapd-test
DEBUG: ACTION: creating /var/run/dirsrv
DEBUG: CMD: systemctl enable dirsrv@test ; STDOUT:  ; STDERR: Created symlink /etc/systemd/system/multi-user.target.wants/dirsrv@test.service → /usr/lib/systemd/system/dirsrv@.service.

DEBUG: ACTION: Creating certificate database is /etc/dirsrv/slapd-test
DEBUG: Allocate <class 'lib389.DirSrv'> with None
DEBUG: Allocate <class 'lib389.DirSrv'> with m2.example.test:1389
DEBUG: Allocate <class 'lib389.DirSrv'> with m2.example.test:1389
DEBUG: nss cmd: /usr/bin/certutil -N -d /etc/dirsrv/slapd-test -f /etc/dirsrv/slapd-test/pwdfile.txt
DEBUG: nss output:
DEBUG: nss cmd: /usr/bin/certutil -L -n Self-Signed-CA -d /etc/dirsrv/ssca/
DEBUG: CSR subject -> CN=m2.example.test,givenName=fb995d04-c4e4-47bf-a160-30b371b57eb0,O=testing,L=389ds,ST=Queensland,C=AU
DEBUG: CSR alt_names -> ['m2.example.test']
DEBUG: nss cmd: /usr/bin/certutil -R --keyUsage digitalSignature,nonRepudiation,keyEncipherment,dataEncipherment --nsCertType sslClient,sslServer --extKeyUsage clientAuth,serverAuth -s CN=m2.example.test,givenName=fb995d04-c4e4-47bf-a160-30b371b57eb0,O=testing,L=389ds,ST=Queensland,C=AU -8 m2.example.test -g 4096 -d /etc/dirsrv/slapd-test -z /etc/dirsrv/slapd-test/noise.txt -f /etc/dirsrv/slapd-test/pwdfile.txt -a -o /etc/dirsrv/slapd-test/Server-Cert.csr
DEBUG: nss cmd: /usr/bin/certutil -C -d /etc/dirsrv/ssca/ -f /etc/dirsrv/ssca//pwdfile.txt -v 24 -a -i /etc/dirsrv/slapd-test/Server-Cert.csr -o /etc/dirsrv/slapd-test/Server-Cert.crt -c Self-Signed-CA
DEBUG: nss cmd: /usr/bin/c_rehash /etc/dirsrv/slapd-test
DEBUG: nss cmd: /usr/bin/certutil -A -n Self-Signed-CA -t CT,, -a -i /etc/dirsrv/slapd-test/ca.crt -d /etc/dirsrv/slapd-test -f /etc/dirsrv/slapd-test/pwdfile.txt
DEBUG: nss cmd: /usr/bin/certutil -A -n Server-Cert -t ,, -a -i /etc/dirsrv/slapd-test/Server-Cert.crt -d /etc/dirsrv/slapd-test -f /etc/dirsrv/slapd-test/pwdfile.txt
DEBUG: nss cmd: /usr/bin/certutil -V -d /etc/dirsrv/slapd-test -n Server-Cert -u YCV
DEBUG: systemd status -> True
DEBUG: systemd status -> True
Job for dirsrv@test.service failed because the control process exited with error code.
See "systemctl status dirsrv@test.service" and "journalctl -xe" for details.
DEBUG: Command '['systemctl', 'start', 'dirsrv@test']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/usr/sbin/dscreate", line 78, in <module>
    result = args.func(inst, log, args)
  File "/usr/lib/python3.6/site-packages/lib389/cli_ctl/instance.py", line 68, in instance_create
    if sd.create_from_inf(args.file):
  File "/usr/lib/python3.6/site-packages/lib389/instance/setup.py", line 533, in create_from_inf
    self.create_from_args(general, slapd, backends, self.extra)
  File "/usr/lib/python3.6/site-packages/lib389/instance/setup.py", line 669, in create_from_args
    self._install_ds(general, slapd, backends)
  File "/usr/lib/python3.6/site-packages/lib389/instance/setup.py", line 888, in _install_ds
    ds_instance.start(timeout=60)
  File "/usr/lib/python3.6/site-packages/lib389/__init__.py", line 1129, in start
    "dirsrv@%s" % self.serverid])
  File "/usr/lib64/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['systemctl', 'start', 'dirsrv@test']' returned non-zero exit status 1.
ERROR: Error: Command '['systemctl', 'start', 'dirsrv@test']' returned non-zero exit status 1.

the dse.ldif got the custom path:
grep testdata /etc/dirsrv/slapd-test/dse.ldif
nsslapd-directory: /testdata/dirsrv/test/db
nsslapd-db-logdirectory: /testdata/dirsrv/test/db

less /var/log/dirsrv/slapd-test/errors
        389-Directory/1.4.2.12 B2021.029.2040
        m2.example.test:1389 (/etc/dirsrv/slapd-test)

[02/Feb/2021:17:54:42.454458825 -0800] - INFO - main - 389-Directory/1.4.2.12 B2021.029.2040 starting up
[02/Feb/2021:17:54:42.457603853 -0800] - INFO - main - Setting the maximum file descriptor limit to: 262144
[02/Feb/2021:17:54:43.008425092 -0800] - INFO - PBKDF2_SHA256 - Based on CPU performance, chose 2048 rounds
[02/Feb/2021:17:54:43.015806960 -0800] - WARN - spal_meminfo_get - cgroups v1 or v2 unable to be read - may not be on this platform ...
[02/Feb/2021:17:54:43.019459291 -0800] - INFO - bdb_config_upgrade_dse_info - create config entry from old config
[02/Feb/2021:17:54:43.027048488 -0800] - WARN - spal_meminfo_get - cgroups v1 or v2 unable to be read - may not be on this platform ...
[02/Feb/2021:17:54:43.030485385 -0800] - NOTICE - ldbm_back_start - found 1870824k physical memory
[02/Feb/2021:17:54:43.033408626 -0800] - NOTICE - ldbm_back_start - found 1115964k available
[02/Feb/2021:17:54:43.036323091 -0800] - NOTICE - ldbm_back_start - cache autosizing: db cache: 46770k
[02/Feb/2021:17:54:43.039575161 -0800] - WARN - spal_meminfo_get - cgroups v1 or v2 unable to be read - may not be on this platform ...
[02/Feb/2021:17:54:43.043020082 -0800] - NOTICE - ldbm_back_start - total cache size: 38314475 B;
[02/Feb/2021:17:54:43.046258626 -0800] - ERR - bdb_version_write - Could not open file "%s" for writing Netscape Portable Runtime %d (%s)
 - /testdata/dirsrv/test/db/DBVERSION[02/Feb/2021:17:54:43.049417611 -0800] - ERR - mkdir_p - /testdata/dirsrv: error -5966 (Access Denied.)
[02/Feb/2021:17:54:43.052351412 -0800] - CRIT - bdb_start - Can't start because the database directory "/testdata/dirsrv/test/db" either doesn't exist, or is not accessible
[02/Feb/2021:17:54:43.055624040 -0800] - ERR - ldbm_back_start - Failed to init database, err=-1 BDB0092 Unknown error: -1
[02/Feb/2021:17:54:43.058984644 -0800] - ERR - plugin_dependency_startall - Failed to start database plugin ldbm database
[02/Feb/2021:17:54:43.066593297 -0800] - CRIT - dblayer_setup - dblayer_init failed
[02/Feb/2021:17:54:43.070416830 -0800] - ERR - ldbm_back_start - Failed to setup dblayer
[02/Feb/2021:17:54:43.073265352 -0800] - ERR - plugin_dependency_startall - Failed to start database plugin ldbm database
[02/Feb/2021:17:54:43.076862239 -0800] - ERR - plugin_dependency_startall - Failed to resolve plugin dependencies
[02/Feb/2021:17:54:43.079920654 -0800] - ERR - plugin_dependency_startall - object plugin Roles Plugin is not started
[02/Feb/2021:17:54:43.082900803 -0800] - ERR - plugin_dependency_startall - accesscontrol plugin ACL Plugin is not started
[02/Feb/2021:17:54:43.085902582 -0800] - ERR - plugin_dependency_startall - preoperation plugin ACL preoperation is not started
[02/Feb/2021:17:54:43.089347403 -0800] - ERR - plugin_dependency_startall - object plugin Class of Service is not started
[02/Feb/2021:17:54:43.092701535 -0800] - ERR - plugin_dependency_startall - object plugin Views is not started
[02/Feb/2021:17:54:43.095761713 -0800] - ERR - plugin_dependency_startall - betxnpreoperation plugin 7-bit check is not started
[02/Feb/2021:17:54:43.098684284 -0800] - ERR - plugin_dependency_startall - preoperation plugin Account Usability Plugin is not started
[02/Feb/2021:17:54:43.102333339 -0800] - ERR - plugin_dependency_startall - betxnpreoperation plugin Auto Membership Plugin is not started
[02/Feb/2021:17:54:43.105615976 -0800] - ERR - plugin_dependency_startall - preoperation plugin deref is not started
[02/Feb/2021:17:54:43.109015111 -0800] - ERR - plugin_dependency_startall - preoperation plugin HTTP Client is not started
[02/Feb/2021:17:54:43.112206577 -0800] - ERR - plugin_dependency_startall - betxnpreoperation plugin Linked Attributes is not started
[02/Feb/2021:17:54:43.115890468 -0800] - ERR - plugin_dependency_startall - betxnpreoperation plugin Managed Entries is not started
[02/Feb/2021:17:54:43.119210915 -0800] - ERR - plugin_dependency_startall - object plugin Multimaster Replication Plugin is not started
[02/Feb/2021:17:54:43.122747970 -0800] - ERR - plugin_dependency_startall - extendedop plugin whoami is not started
[02/Feb/2021:17:54:43.125953111 -0800] - ERR - plugin_dependency_startall - database plugin ldbm database is not started
(END)

Expected results:
yes

Additional info:

got incorrect sub directory permissions after the dscreate, 770 instead of 775, in this example with /testdata/ and /testdata/dirsrv/

ls -laRZ /testdata/
/testdata/:
total 4
drwxrwx---.  3 root root unconfined_u:object_r:default_t:s0   20 Feb  2 19:55 .
dr-xr-xr-x. 20 root root system_u:object_r:root_t:s0        4096 Feb  2 19:55 ..
drwxrwx---.  3 root root unconfined_u:object_r:default_t:s0   18 Feb  2 19:55 dirsrv

/testdata/dirsrv:
total 0
drwxrwx---. 3 root      root       unconfined_u:object_r:default_t:s0 18 Feb  2 19:55 .
drwxrwx---. 3 root      root       unconfined_u:object_r:default_t:s0 20 Feb  2 19:55 ..
drwxrwx---. 3 ldapuser1 ldapgroup1 unconfined_u:object_r:default_t:s0 16 Feb  2 19:55 test

/testdata/dirsrv/test:
total 0
drwxrwx---. 3 ldapuser1 ldapgroup1 unconfined_u:object_r:default_t:s0 16 Feb  2 19:55 .
drwxrwx---. 3 root      root       unconfined_u:object_r:default_t:s0 18 Feb  2 19:55 ..
drwxrwx---. 2 ldapuser1 ldapgroup1 unconfined_u:object_r:default_t:s0  6 Feb  2 19:55 db

/testdata/dirsrv/test/db:
total 0
drwxrwx---. 2 ldapuser1 ldapgroup1 unconfined_u:object_r:default_t:s0  6 Feb  2 19:55 .
drwxrwx---. 3 ldapuser1 ldapgroup1 unconfined_u:object_r:default_t:s0 16 Feb  2 19:55 ..

### PR DESCRIPTION
**Pull request by progier389**
Problem:
When creating an instance with a custom dir_path and if seLinux is enabled then the instance creation fails:
Instance fails to start because database cannot be created due to seLinux permission error.
    
Solution:
The solution is to set the right seLinux label for the instance directories by deleting current context 
and adding new one 
but there are some pitfalls to avoid:
   - Make sure that the right path is used (otherwise semanage fails)
   - Make sure that we do not explicity try to overrule a record explicity set by the seLinux policy
     ( for example trying to set db_dir as /tmp ( and tmp_dir as something else )
The idea is to parse the semanage fcontext -l [-C] output to get the:
    path equivalence rules (about symbolic link)
    local file context path and labels
    global policy file context path and labels
And use them to compute the right path, and to decide what must be done 
  (nothing, delete existing local context, raise an error or add new context)

Also made sure that the selinux labels stored in the local policy get cleaned
  when removing and instance (and that all labels set by ds get cleaned when removing all instances)  
    
Note: also fixed some lib389 warning about obsolete escape sequences
 and also removed _deleteDirsrv private function (dead code) to avoid that 
 someone else waste time trying to "fix" it.
    
Issue: [4592](https://github.com/389ds/389-ds-base/issues/4592)
    
Reviewd by:

### CODE REVIEW

File: dirsrvtests/tests/suites/setup_ds/dscreate_test.py [hunk 1]
```diff
@@ -168,6 +169,79 @@ def test_setup_ds_minimal(topology):
...
 
+
+@pytest.mark.skipif(os.getuid()!=0, reason="pytest non run by root user")
+def test_setup_ds_custom_db_dir(topology):
+    """Test DS setup using custom uid,gid and db_dir path
+
+    :id: 5a596887-cabb-4862-a91c-5eedafe222cd
+    :setup: standalone instance
+    :steps:
+        1. Create the user that will run ns-slapd
```
Comment (progier389) [inline dirsrvtests/tests/suites/setup_ds/dscreate_test.py:201]: I forgot to renumber the steps ! -;)

Comment (progier389) [reply dirsrvtests/tests/suites/setup_ds/dscreate_test.py:201]: Fixed

**Comment by droideck** [comment]
Hmm, when I run locally on a Fedora 36 machine, it fails for me to start the server. The logs say:

```
[30/Aug/2022:13:06:13.581028543 -0400] - ERR - libdb - /mydirsrv/db: Permission denied
[30/Aug/2022:13:06:13.584245966 -0400] - ERR - libdb - BDB0061 PANIC: Permission denied
[30/Aug/2022:13:06:13.607596039 -0400] - ERR - libdb - BDB1546 unable to join the environment
[30/Aug/2022:13:06:13.641692000 -0400] - CRIT - bdb_start - Opening database environment (/dev/shm/slapd-standalone) failed. err=-30973: Unexpected dbimpl error code
[30/Aug/2022:13:06:13.644570069 -0400] - ERR - ldbm_back_start - Failed to init database, err=-30973 Unexpected dbimpl error code
[30/Aug/2022:13:06:13.647110339 -0400] - ERR - plugin_dependency_startall - Failed to start database plugin ldbm database
[30/Aug/2022:13:06:13.653402699 -0400] - CRIT - dblayer_setup - dblayer_init failed
[30/Aug/2022:13:06:13.656307990 -0400] - ERR - ldbm_back_start - Failed to setup dblayer
[30/Aug/2022:13:06:13.658774814 -0400] - ERR - plugin_dependency_startall - Failed to start database plugin ldbm database
[30/Aug/2022:13:06:13.661145546 -0400] - ERR - plugin_dependency_startall - Failed to resolve plugin dependencies
[30/Aug/2022:13:06:13.663582611 -0400] - ERR - plugin_dependency_startall - object plugin Roles Plugin is not started
[30/Aug/2022:13:06:13.665788979 -0400] - ERR - plugin_dependency_startall - accesscontrol plugin ACL Plugin is not started
[30/Aug/2022:13:06:13.668145714 -0400] - ERR - plugin_dependency_startall - preoperation plugin ACL preoperation is not started
[30/Aug/2022:13:06:13.674676226 -0400] - ERR - plugin_dependency_startall - object plugin Class of Service is not started
[30/Aug/2022:13:06:13.677828839 -0400] - ERR - plugin_dependency_startall - object plugin Views is not started
[30/Aug/2022:13:06:13.680297227 -0400] - ERR - plugin_dependency_startall - betxnpreoperation plugin 7-bit check is not started
[30/Aug/2022:13:06:13.704856042 -0400] - ERR - plugin_dependency_startall - preoperation plugin Account Usability Plugin is not started
[30/Aug/2022:13:06:13.707787827 -0400] - ERR - plugin_dependency_startall - betxnpreoperation plugin Auto Membership Plugin is not started
[30/Aug/2022:13:06:13.710421502 -0400] - ERR - plugin_dependency_startall - preoperation plugin deref is not started
[30/Aug/2022:13:06:13.712868457 -0400] - ERR - plugin_dependency_startall - betxnpreoperation plugin Linked Attributes is not started
[30/Aug/2022:13:06:13.714940257 -0400] - ERR - plugin_dependency_startall - betxnpreoperation plugin Managed Entries is not started
[30/Aug/2022:13:06:13.717200120 -0400] - ERR - plugin_dependency_startall - object plugin Multisupplier Replication Plugin is not started
[30/Aug/2022:13:06:13.719523888 -0400] - ERR - plugin_dependency_startall - extendedop plugin whoami is not started
[30/Aug/2022:13:06:13.722040779 -0400] - ERR - plugin_dependency_startall - database plugin ldbm database is not started
```

**Comment by progier389** [comment]
[30/Aug/2022:13:06:13.581028543 -0400] - ERR - libdb - /mydirsrv/db: Permission denied
Is typically the error that you get without the fix. 
I suspect that **ls -ldZ /mydirsrv/db**  shows the wrong label. (It should be dirsrv_var_lib_t).
IMHO You should do the manual test (using the following commands and compare the result
  I suspect that a python module is missing and dscreate should log an error
   (BTW Can you check that **semanage** command is present (i.e policycoreutils-python-utils package should be installed),
     if that is not the case then you may get trouble even without this fix (because it is also used for port relabeling)) 

Anyway when I done the test I saw a lot of _libsemanage.dbase_llist_query: could not query record value_ warnings
 so I guess I should first check for record existence before querying it to avoid this noise ...

```
[root@linux 389-ds-base]# dsctl  standalone  remove --do-it
Removing instance ...
Completed instance removal

[root@linux 389-ds-base]# echo "
[general]
[slapd]
db_dir = /mydirsrv/db
instance_name = standalone
[backend-userroot]
sample_entries = yes
suffix = dc=example,dc=com" >/tmp/t

[root@linux 389-ds-base]# dscreate from-file /tmp/t
Starting installation ...
Validate installation settings ...
Create file system structures ...
Create self-signed certificate database ...
Perform SELinux labeling ...
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
setting selinux type dirsrv_var_run_t in file context /var/run/dirsrv
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
setting selinux type dirsrv_var_lock_t in file context /var/run/lock/dirsrv/slapd-standalone
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
libsemanage.dbase_llist_query: could not query record value
Create database backend: dc=example,dc=com ...
Perform post-installation tasks ...
Completed installation for instance: slapd-standalone

[root@linux 389-ds-base]# ls -lZd  /mydirsrv/db
drwxrwx---. 3 dirsrv dirsrv unconfined_u:object_r:dirsrv_var_lib_t:s0 4096 Aug 31 15:59 /mydirsrv/db

```

**Comment by progier389** [comment]
@droideck, Doing my tests, I saw another issue (which could be your problem) 
   as dsctl remove does not delete the dir_path, if you have run the test with a non fixed version, you need to perform a 
     rm -rf /mydirsrv to insure that no files remain with a wrong label (apparently restorecon (with recursive flags) 
       does not reset the label)
   I can do the rm in the test case but I do not really know how to prevent the issue in real usage 
     (removing the files in the path in dscreate seems dangerous ... )

**Comment by droideck** [comment]
> @droideck, Doing my tests, I saw another issue (which could be your problem) as dsctl remove does not delete the dir_path, if you have run the test with a non fixed version, you need to perform a rm -rf /mydirsrv to insure that no files remain with a wrong label (apparently restorecon (with recursive flags) does not reset the label) I can do the rm in the test case but I do not really know how to prevent the issue in real usage (removing the files in the path in dscreate seems dangerous ... )

Hmm, can we use some other directory for non-prefix installations? (may be in /tmp/mydirsrv/db if it's allowed)
Also, please make sure that you remove the instance after the test is finished (with 'finally' or 'request.addfinalizer()')

I've tested the fixed version though. It was built freshly on a separate virtual machine.

I've run the test once again with DEBUGGING=yes and checked 'errors' log.
It had this:

```
[31/Aug/2022:21:41:37.050717922 -0400] - ERR - libdb - /mydirsrv/db: Permission denied
[31/Aug/2022:21:41:37.061161333 -0400] - ERR - libdb - BDB0061 PANIC: Permission denied
[31/Aug/2022:21:41:37.080127204 -0400] - ERR - libdb - BDB1546 unable to join the environment
[31/Aug/2022:21:41:37.106234689 -0400] - CRIT - bdb_start - Opening database environment (/dev/shm/slapd-standalone) failed. err=-30973: Unexpected dbimpl error code
[31/Aug/2022:21:41:37.109988274 -0400] - ERR - ldbm_back_start - Failed to init database, err=-30973 Unexpected dbimpl error code
[31/Aug/2022:21:41:37.112653888 -0400] - ERR - plugin_dependency_startall - Failed to start database plugin ldbm database
[31/Aug/2022:21:41:37.118270986 -0400] - CRIT - dblayer_setup - dblayer_init failed
[31/Aug/2022:21:41:37.122090699 -0400] - ERR - ldbm_back_start - Failed to setup dblayer
```

Also, the new directory didn't have proper context:

```
[root@mymachine ds]# ll -Z /mydirsrv/
total 8
drwxrwx---. 2 ldapsrv1 ldapsrv1 unconfined_u:object_r:default_t:s0 4096 Aug 31 21:41 db
```

**Comment by progier389** [comment]
Ok, I finally was able to also get the issue. 
   Apparently using directly the lib is not a good idea and leads to inconsistencies (I knew that It was a bad idea ! -;) ). 
   The good news is that I was able able to parse semanage fcontext -l [-C] output to get the all info I need to compute the path and determine if label must be deleted or not. so I get rid of the library. 
 Note: about the test the instance was already deleted after the test, I just added code to insure that custom directory get deleted and its label removed (both before (to avoid trouble in DEBUG mode) and after the test) and also added a check to make sure that semanage is installed. 
  Lets hope that test work in your environment.

**Comment by progier389** [comment]
BTW I cannot put db_dir under /tmp or /var/tmp (get a fatal error while trying to access the filesystem info:
ERR - bdb_no_diskspace - Cannot get file system info for (/tmp/pytest_test_setup_ds_custom_db_dir/db); file system corrupted?
 So I used: /dirsrv_pytest_test_setup_ds_custom_db_dir/db ( with a more meaningful name)

**Comment by progier389** [comment]
Needed a few tries but now the tests are OK.

**Comment by progier389** [comment]
Fixed an issue in pytest while running on 1minutesTips VM, because selinux_fcontext_info cache was not updated when deleting or adding a label ( so if file label was changed twice, the second calls use wrong cache info and fails )

**Comment by progier389** [comment]
Added a test to check that selinux label are removed after the instances deletion 
and realized that I fixed the wrong function  ( _deleteDirsrv private function is dead code so I removed it to avoid further confusion and fixed   lib389.instance.remove.py )

**Review by droideck — APPROVED** [review]
Looks good to me!

**Pull request merged**